File: plugins/rider/src/main/kotlin/com/shortsleevestudio/gamescript/handlers/MessageMediator.kt
package com.shortsleevestudio.gamescript.handlers

import com.google.gson.JsonObject
import com.intellij.openapi.diagnostic.Logger

/**
 * Routes incoming messages to their appropriate handlers.
 * Each handler is registered by message type string.
 */
class MessageMediator {

    private val handlers = mutableMapOf<String, suspend (JsonObject) -> Unit>()

    /**
     * Register a handler for a specific message type.
     */
    fun register(type: String, handler: suspend (JsonObject) -> Unit) {
        handlers[type] = handler
    }

    /**
     * Register multiple handlers at once.
     */
    fun registerAll(handlersMap: Map<String, suspend (JsonObject) -> Unit>) {
        handlers.putAll(handlersMap)
    }

    /**
     * Handle an incoming message by routing to the appropriate handler.
     * Re-throws exceptions after logging for caller to handle.
     */
    suspend fun handle(type: String, message: JsonObject) {
        val handler = handlers[type]
        if (handler != null) {
            try {
                handler(message)
            } catch (e: Exception) {
                LOG.error("Handler for '$type' threw exception", e)
                throw e
            }
        } else {
            LOG.warn("No handler registered for message type: $type")
        }
    }

    /**
     * Check if a handler is registered for a message type.
     */
    fun hasHandler(type: String): Boolean = handlers.containsKey(type)

    /**
     * Get all registered message types.
     */
    fun getRegisteredTypes(): Set<String> = handlers.keys.toSet()

    companion object {
        private val LOG = Logger.getInstance(MessageMediator::class.java)
    }
}

File: plugins/dotnet/visual-studio/GameScript.VisualStudio/Handlers/MessageMediator.cs
using System.Text.Json;

namespace GameScript.VisualStudio;

/// <summary>
/// Delegate for message handlers.
/// </summary>
public delegate Task MessageHandlerDelegate(JsonElement message, CancellationToken cancellationToken);

/// <summary>
/// Routes incoming messages to their appropriate handlers.
/// Each handler is registered by message type string.
/// </summary>
public sealed class MessageMediator
{
    private readonly Dictionary<string, MessageHandlerDelegate> _handlers = new();

    /// <summary>
    /// Register a handler for a specific message type.
    /// </summary>
    public void Register(string type, MessageHandlerDelegate handler)
    {
        if (_handlers.ContainsKey(type))
        {
            System.Diagnostics.Debug.WriteLine($"[MessageMediator] Handler for '{type}' already registered, overwriting");
        }
        _handlers[type] = handler;
    }

    /// <summary>
    /// Register multiple handlers at once.
    /// </summary>
    public void RegisterMany(IEnumerable<KeyValuePair<string, MessageHandlerDelegate>> handlers)
    {
        foreach (var (type, handler) in handlers)
        {
            Register(type, handler);
        }
    }

    /// <summary>
    /// Handle an incoming message by routing to the appropriate handler.
    /// </summary>
    public async Task HandleAsync(string type, JsonElement message, CancellationToken cancellationToken)
    {
        if (_handlers.TryGetValue(type, out var handler))
        {
            try
            {
                await handler(message, cancellationToken);
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"[MessageMediator] Handler for '{type}' threw: {ex.Message}");
                throw;
            }
        }
        else
        {
            System.Diagnostics.Debug.WriteLine($"[MessageMediator] No handler registered for message type: {type}");
        }
    }

    /// <summary>
    /// Check if a handler is registered for a message type.
    /// </summary>
    public bool HasHandler(string type) => _handlers.ContainsKey(type);

    /// <summary>
    /// Get all registered message types.
    /// </summary>
    public IEnumerable<string> GetRegisteredTypes() => _handlers.Keys;
}
